### PR TITLE
This adds the ability to log memory changes to vcd output

### PR DESCRIPTION
**Pull request by chick**
Usage
  --tr-mem-to-vcd commands

Commands are of the form

```
all.               all indicies for all memories
all:h              all indices for all memories, indices rendered in hex
mem1:all           all indices for memory1
mem1:1,2,3,10-20    mem1 will only have listed indices logged
mem1:0x3,0x10-0x20  mem1 will only have listed indices logged, indices will be rendered in hex
```

tr-mem-to-vcd can appear multiple times on command line to specify other memories

MemoryToVCD annotations can be used instead of command line

- Creates a new VcdMemoryLoggingController that can track what should be logged
  - All memories and all indices can be specified
  - Specific memories and all indices can be specified
  - specific memories and specific ranges of indices can be specified
  - indices can be specified in a radix and that radix will be used for vcd serialization of memory index
- Some additional tools for handling of string to indices were collected in NumberHelpers
  - This was derived from some code in TreadleRepl which is now change to use the new object
- New options to control this logging
  - default is same as previous (no logging of memories)
  - --tr-mem-to-vcd "loggingcommand"
  - MemoryToVCD(loggingCommand)
  - multiple occurrences of command line flag and annotation are supported
- Records memory locations being logged to vcd directory
- Testing specification is provided

### CODE REVIEW

**Comment by ekiwi-sifive** [comment]
I can review this PR this week. Can you add me (@ekiwi-sifive) as a reviewer?

File: src/main/scala/treadle/TreadleOptions.scala [hunk 1]
```diff
@@ -162,6 +162,23 @@ case object RollBackBuffersAnnotation extends HasShellOptions {
   )
 }
 
+/**
+  *  Sets the number of rollback buffers in simulator, useful to see why wires have their values
```
Comment (ekiwi-sifive) [inline src/main/scala/treadle/TreadleOptions.scala:166]: This comment seems to be a copy-and-paste error.

File: src/main/scala/treadle/executable/VcdMemoryLoggingController.scala [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+  * if logAllRadixOpt is defined then all indices for all memories should be logged
+  * otherwise if info has an entry then check the memory index
+  * if the set contains -1 then all indices should be logged for that memory
+  *
+  * @param logAllRadixOpt  if defined then log all memories using this radix
+  * @param memoriesTracked    map of memories to the set of indices that should be logged
+  */
+class VcdMemoryLoggingController(
+  logAllRadixOpt:  Option[Int] = None,
+  memoriesTracked: Map[Symbol, IndicesAndRadix] = Map.empty
```
Comment (ekiwi-sifive) [inline src/main/scala/treadle/executable/VcdMemoryLoggingController.scala:35]: is `memoriesTracked` ignored when `logAllRadixOpt` is `Some`?
This seems to be the case in `vcdKey` but not in `getIndexedName`. Why?

File: src/main/scala/treadle/executable/VcdMemoryLoggingController.scala [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+) {
+
+  private def indexedName(name: String, index: Int, radix: Int): String = {
+    s"$name(${BigInt(index).toString(radix)})"
+  }
+
+  /* generate a vcd element name for a given memory location
```
Comment (ekiwi-sifive) [inline src/main/scala/treadle/executable/VcdMemoryLoggingController.scala:42]: `/**`

File: src/main/scala/treadle/executable/VcdMemoryLoggingController.scala [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+              Some(indexedName(symbol.name, offset, radix))
+            } else {
+              None
+            }
+          case _ => None
+        }
+    }
+  }
+
+  /* Builds a list of all tracked memories and the locations within them that are tracked
```
Comment (ekiwi-sifive) [inline src/main/scala/treadle/executable/VcdMemoryLoggingController.scala:62]: `/**`

File: src/main/scala/treadle/utils/NumberHelpers.scala [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+limitations under the License.
+ */
+
+package treadle.utils
+
+import treadle.executable.{IndicesAndRadix, TreadleException}
+
+import scala.util.matching.Regex
+
+object NumberHelpers {
```
Comment (ekiwi-sifive) [inline src/main/scala/treadle/utils/NumberHelpers.scala:23]: Could you write some unit tests for the functions in this object?

File: src/main/scala/treadle/TreadleOptions.scala (hunk 1 above)
Comment (chick) [reply src/main/scala/treadle/TreadleOptions.scala:166]: Yep, cut and paste error. Fixed.

File: src/main/scala/treadle/executable/VcdMemoryLoggingController.scala (hunk 2 above)
Comment (chick) [reply src/main/scala/treadle/executable/VcdMemoryLoggingController.scala:35]: Great catch. This is only used for building the directory in the VCD header.

File: src/main/scala/treadle/executable/VcdMemoryLoggingController.scala (hunk 4 above)
Comment (chick) [reply src/main/scala/treadle/executable/VcdMemoryLoggingController.scala:62]: Done

**Comment by chick** [comment]
@ekiwi-sifive I think I have addressed all your change requests

**Comment by ekiwi-sifive** [comment]
>  I think I have addressed all your change requests

Thanks!

🚢

**Comment by ekiwi-sifive** [comment]
@chick can you merge it yourself? I am not sure what the correct convention for treadle is (merge, squash or rebase).

**Comment by chick** [comment]
@ekiwi-sifive Yes, I can, I'm happy to do the merge for any PR here after it's been approved.

**Pull request merged**